src/validation.cpp

```diff
@@ -1090,6 +1090,33 @@ bool AcceptToMemoryPool(CTxMemPool& pool, TxValidationState &state, const CTrans
     return AcceptToMemoryPoolWithTime(chainparams, pool, state, tx, GetTime(), plTxnReplaced, bypass_limits, test_accept, fee_out);
 }
 
+// Calculate the fee of a transaction by looking up inputs from the mempool if
+// necessary. Return true if the fee was successfully calculated, or false if
+// an error with the transaction was discovered.
+// If the transaction is valid, then the fee returned is correct.
+// However, success does not imply that the transaction is valid.
+bool GetTransactionFee(CTxMemPool& pool, TxValidationState &state, const CTransactionRef &tx, CAmount *fee_out)
+{
+    LOCK2(cs_main, pool.cs);
+
+    CCoinsViewMemPool viewmempool(&::ChainstateActive().CoinsTip(), pool);
+    CCoinsViewCache view(&viewmempool);
+    view.GetBestBlock();
+
+    *fee_out = 0;
+
+    // Fees don't make sense for coinbase transactions
+    if (tx->IsCoinBase()) {
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "coinbase");
```

Comment:
It feels weird to be setting a `TxValidationResult` in a function that purports to simply be a utility helper, and duplicating validity checks in this function feels like code duplication with the risk of divergence in the future. I'd much prefer it if this function was only checking/returning the fee, and all validity state was handled by ATMP.